File: convert.py
from configparser import ConfigParser
from pathlib import Path, PureWindowsPath
import shutil
from win2xcur.parser import open_blob
from win2xcur.writer import to_smart

cursors = [
    "default.cur", # Normal Select
    "help.cur", # Help Select
    "progress.ani", # Working in Background
    "wait.ani", # Busy
    "cross.cur", # Precision Select
    "text.cur", # Text Select
    "pencil.cur", # Handwriting
    "not-allowed.cur", # Unavailable
    "ns-resize.cur", # Vertical Resize
    "ew-resize.cur", # Horizontal Resize
    "nwse-resize.cur", # Diagonal Resize 1
    "nesw-resize.cur", # Diagonal Resize 2
    "all-scroll.cur", # Move
    "up-arrow.cur", # Alternate Select
    "pointer.cur", # Link Select
    "", # Location Select
    "", # Person Select
]

def processFolder(baseDir):
    if not baseDir.exists():
        raise "Folder not found"

    schemeName = Path(baseDir).name.split("-")[0:-1];
    schemeName = " ".join(schemeName)
    inDir = baseDir / "cursors";
    outDir = baseDir / "win";

    if outDir.exists():
        shutil.rmtree(outDir)
    outDir.mkdir()

    inFiles = list(filter(bool, cursors))
    inFiles = [Path(v).stem for v in inFiles]
    inFiles = [inDir / v for v in inFiles]

    for file in inFiles:
        blob = file.open('rb').read()
        cursor = open_blob(blob)
        ext, result = to_smart(cursor.frames)
        output = outDir / (file.name + ext)
        with open(output, 'wb') as f:
            f.write(result)

    cursorDir = PureWindowsPath("Cursors").joinpath(*schemeName.split(" "))
    regCursors = [v and PureWindowsPath("%SYSTEMROOT%") / cursorDir / v for v in cursors]
    regCursors = ",".join([str(v) for v in regCursors])

    schemeCur = list(filter(bool, cursors))
    schemeCur = {i: v for i, v in enumerate(schemeCur)}

    inf = ConfigParser(interpolation = None)
    inf.read_dict({
        "Version": { "signature": "$CHICAGO$" },
        "DefaultInstall": {
            "CopyFiles": "Scheme.Cur, Scheme.Txt",
            "AddReg": "Scheme.Reg",
        },
        "DestinationDirs": {
            "Scheme.Cur": f'10,"{cursorDir}"',
            "Scheme.Txt": f'10,"{cursorDir}"',
        },
        "Scheme.Reg": {
           0: f'HKCU,"Control Panel\\Cursors\\Schemes","{schemeName}",0x00020000,"{regCursors}"',
        },
        "Scheme.Cur": schemeCur
    })
    with open(outDir / "install.inf", "w") as configfile:
        inf.write(configfile)

if __name__ == "__main__":
    dirs = list(Path("cursors").iterdir())
    dirs = [v for v in dirs if v.is_dir()]
    for dir in dirs:
        name = dir.name.replace('-', ' ')
        print("Converting", name)
        processFolder(dir)

File: convert.ts
#!/usr/bin/env -S deno run -A

import $, { PathRef } from "https://deno.land/x/dax@0.35.0/mod.ts";
import * as ini from "https://deno.land/std@0.209.0/ini/mod.ts";
import * as path from "https://deno.land/std@0.209.0/path/windows/mod.ts";

// $.setPrintCommand(true);

const cursors = [
  "default.cur", // Normal Select
  "help.cur", // Help Select
  "progress.ani", // Working in Background
  "wait.ani", // Busy
  "cross.cur", // Precision Select
  "text.cur", // Text Select
  "pencil.cur", // Handwriting
  "not-allowed.cur", // Unavailable
  "ns-resize.cur", // Vertical Resize
  "ew-resize.cur", // Horizontal Resize
  "nwse-resize.cur", // Diagonal Resize 1
  "nesw-resize.cur", // Diagonal Resize 2
  "all-scroll.cur", // Move
  "up-arrow.cur", // Alternate Select
  "pointer.cur", // Link Select
  "", // Location Select
  "", // Person Select
];

async function processFolder(baseDir: PathRef | string) {
  baseDir = $.path(baseDir);

  if (!(await baseDir.exists())) throw new Deno.errors.NotFound();

  const schemeName = baseDir.basename().split("-").slice(0, -1).join(" ");
  const inDir = baseDir.join("cursors");
  const outDir = baseDir.join("win");
  await outDir.emptyDir();

  const inFiles = cursors
    .filter(Boolean)
    .map((v) => path.basename(v, path.extname(v)))
    .map((v) => inDir.join(v));
  await $`python -m pipx run x2wincur ${inFiles} -o ${outDir}`;

  const cursorDir = path.join("Cursors", ...schemeName.split(" "));
  const regCursors = cursors
    .map((v) => v && path.join("%SYSTEMROOT%", cursorDir, v))
    .join(",");

  const inf = {
    Version: { signature: "$CHICAGO$" },
    DefaultInstall: {
      CopyFiles: "Scheme.Cur, Scheme.Txt",
      AddReg: "Scheme.Reg",
    },
    DestinationDirs: {
      "Scheme.Cur": `10,"${cursorDir}"`,
      "Scheme.Txt": `10,"${cursorDir}"`,
    },
    "Scheme.Reg": [
      `HKCU,"Control Panel\\Cursors\\Schemes","${schemeName}",0x00020000,"${regCursors}"`,
    ],
    "Scheme.Cur": cursors.filter(Boolean),
  };

  outDir.join("install.inf").writeText(ini.stringify(inf));
}

if (import.meta.main) {
  const dirs = [...$.path("cursors").readDirSync()]
    .filter((v) => v.isDirectory)
    .map((v) => v.path);

  const pb = $.progress("Converting", {
    length: dirs.length,
  });
  for (const dir of dirs) {
    const name = dir.basename().replaceAll("-", " ");
    pb.message(name);
    // await $.sleep("0.5s");
    await processFolder(dir);
    pb.increment();
  }
  pb.finish();
}
